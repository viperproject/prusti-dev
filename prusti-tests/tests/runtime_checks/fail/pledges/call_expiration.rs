//@run: 101
//@compile-flags: -Pcheck_overflows=false
use prusti_contracts::*;
#[derive(Clone)]
struct Percentage(usize);

#[trusted]
#[insert_runtime_check]
#[requires(p.0 <= 100)]
#[insert_runtime_check]
#[assert_on_expiry(*result <= 100, p.0 <= 100)]
#[insert_runtime_check]
#[ensures(old(p.0) == p.0)]
fn index_mut(p: &mut Percentage) -> &mut usize {
    &mut p.0
}

fn increment(x: &mut usize) {
    *x = *x + 1;
}

fn main() {
    // this one fails because index_mut_other is called
    foo();
}

// This testcase exists, because here the expiration location is a call
// terminator. Right after calling increment(r), r expires.
// This is handled differently than an expiration at an arbitrary statement.
#[trusted]
fn foo() {
    let mut p = Percentage(100);
    let r = index_mut(&mut p);
    increment(r);
    // r expires here, and is too high, so the check fails
    println!("p now has value: {}", p.0);
}

