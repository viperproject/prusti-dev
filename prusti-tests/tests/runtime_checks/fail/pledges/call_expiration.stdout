check function prusti_pre_check_item_index_mut is performed
check function prusti_post_check_item_index_mut is performed
check function prusti_pledge_lhs_check_item_index_mut is performed
